builtin/logical/ssh/storage.go

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package ssh
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openbao/openbao/sdk/v2/helper/errutil"
+	"github.com/openbao/openbao/sdk/v2/logical"
+)
+
+const (
+	defaultRef = "default"
+
+	MigrationInfoKey    = "config/migrationInfo"
+	issuerPrefix        = "config/issuer/"
+	storageIssuerConfig = "config/issuers"
+
+	caPublicKey                       = "ca_public_key"
+	caPrivateKey                      = "ca_private_key"
+	caPublicKeyStoragePath            = "config/ca_public_key"
+	caPublicKeyStoragePathDeprecated  = "public_key"
+	caPrivateKeyStoragePath           = "config/ca_private_key"
+	caPrivateKeyStoragePathDeprecated = "config/ca_bundle"
+
+	maxRolesToScanOnIssuerChange = 100
+	maxRolesToFindOnIssuerChange = 10
+
+	// Used as a quick sanity check for a reference id lookups...
+	uuidLength = 36
+)
+
+const (
+	IssuerRefNotFound = "not-found"
+)
+
+type issuerEntry struct {
+	ID         string `json:"id"`
+	Name       string `json:"name"`
+	PublicKey  string `json:"public_key"`
+	PrivateKey string `json:"private_key"`
+	Version    uint   `json:"version"`
+}
+
+type issuerConfigEntry struct {
+	DefaultIssuerID string `json:"default"`
+}
+
+type storageContext struct {
+	Context context.Context
+	Storage logical.Storage
+	Backend *backend
+}
+
+// makeStorageContext creates a storage context with the given context and storage
+func (b *backend) makeStorageContext(ctx context.Context, s logical.Storage) *storageContext {
+	return &storageContext{
+		Context: ctx,
+		Storage: s,
+		Backend: b,
+	}
+}
+
+// writeIssuer writes an issuerEntry to storage
+func (sc *storageContext) writeIssuer(issuer *issuerEntry) error {
+	issuerId := issuer.ID
+
+	json, err := logical.StorageEntryJSON(issuerPrefix+issuerId, issuer)
+	if err != nil {
+		return err
+	}
+
+	return sc.Storage.Put(sc.Context, json)
+}
+
+// deleteIssuer removes an issuer from storage and unsets from default, if was.
+// Returns a boolean indicating if the issuer was the default issuer
+func (sc *storageContext) deleteIssuer(id string) (bool, error) {
+	config, err := sc.getIssuersConfig()
+	if err != nil {
+		return false, err
+	}
+
+	wasDefault := false
+	if config.DefaultIssuerID == id {
+		wasDefault = true
+		config.DefaultIssuerID = ""
```

Comment:
Just pointing out that there'll be a few places that need to be updated to handle the default being a ref rather than a concrete ID.

Reply:
Given what we discussed [here](https://github.com/openbao/openbao/pull/880#discussion_r1950871123), not anymore I assume?

Reply:
Yep :-)